### PR TITLE
wasm multicore support

### PR DESCRIPTION
**Pull request by prekucki**
This change is [Reviewable](https://reviewable.io/reviews/golemfactory/golem/4889)

### CODE REVIEW

**Review by tworec — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/golemfactory/golem/4889)* status: 0 of 9 files reviewed, 9 unresolved discussions (waiting on @mfranciszkiewicz, @prekucki, @shadeofblue, and @tworec)

___
*[golem/task/taskcomputer.py, line 91 at r1](https://reviewable.io/reviews/golemfactory/golem/4889#-Ltiz_DTElux8xw36AIp:-Ltiz_DTElux8xw36AIq:b-8fx2u7) ([raw file](https://github.com/golemfactory/golem/blob/d243dc3860dc4f0ed2793a9169b1372d833b6821/golem/task/taskcomputer.py#L91)):*
> ```Python
>     def task_given(self, ctd: ComputeTaskDef) -> None:
>         assert not self._new_computer.has_assigned_task()
>         assert self._old_computer.can_take_work()
> ```

:+1: 
___
*[golem/task/taskcomputer.py, line 431 at r2](https://reviewable.io/reviews/golemfactory/golem/4889#-Ltj-geg--g6Tv4KQrOM:-Ltj-geg--g6Tv4KQrON:b7av8yi) ([raw file](https://github.com/golemfactory/golem/blob/6783ae79e57b0013b8ce46534718bf9ea718f05c/golem/task/taskcomputer.py#L431)):*
> ```Python
> 
> @dataclass
> class TaskComputation:
> ```

I'd add a short purpose of this class --> to make multiple computations within single TaskComputer
___
*[golem/task/taskcomputer.py, line 552 at r2](https://reviewable.io/reviews/golemfactory/golem/4889#-LtjMAoQ2-2CVObWekwM:-LtjMAoQ2-2CVObWekwN:b571xb) ([raw file](https://github.com/golemfactory/golem/blob/6783ae79e57b0013b8ce46534718bf9ea718f05c/golem/task/taskcomputer.py#L552)):*
> ```Python
>         return self.task_computer.task_finished(self)
> 
>     def start_computation(self) -> None:  # pylint: disable=too-many-locals
> ```

this is within TaskComputation class so we can use shorter name here eg. `start`
___
*[golem/task/taskcomputer.py, line 648 at r2](https://reviewable.io/reviews/golemfactory/golem/4889#-Ltj5Pyk5-oce45C9eXA:-Ltj5Pyk5-oce45C9eXB:b-kb3374) ([raw file](https://github.com/golemfactory/golem/blob/6783ae79e57b0013b8ce46534718bf9ea718f05c/golem/task/taskcomputer.py#L648)):*
<details><summary><i alt="summary">Quoted 7 lines of code…</i></summary>

> ```
>         # So apparently it is perfectly fine for mypy to assign None to a
> 
>         # non-optional variable. And if I tried Optional['ComputeTaskDef']
> 
>         # then I would get "Optional[Any] is not indexable" error.
> 
>         # Get your sh*t together, mypy!
> ```
</details>

while I like this comment, it is not adequate here anymore, bc you've replaced None with empty list
___
*[golem/task/taskcomputer.py, line 680 at r2](https://reviewable.io/reviews/golemfactory/golem/4889#-LtjA3h3BB_7CyyD_vI4:-Ltj9zEq1FL4CyPzYojB:b-23abis) ([raw file](https://github.com/golemfactory/golem/blob/6783ae79e57b0013b8ce46534718bf9ea718f05c/golem/task/taskcomputer.py#L680)):*
> ```Python
>         if not self.assigned_subtasks:
>             return None
>         return self.assigned_subtasks[0].assigned_task_id
> ```

I'm not sure if such a shortcut is correct. We can potentially have multiple task_id assigned but returning the one for the first subtask. 
`TaskServer.resource_failure` can be possibly affected. This is minor since it affects only error handling
___
*[golem/task/taskcomputer.py, line 718 at r2](https://reviewable.io/reviews/golemfactory/golem/4889#-Ltj76vqAC19-Nxgzrxi:-Ltj76vqAC19-Nxgzrxj:b-fpny89) ([raw file](https://github.com/golemfactory/golem/blob/6783ae79e57b0013b8ce46534718bf9ea718f05c/golem/task/taskcomputer.py#L718)):*
<details><summary><i alt="summary">Quoted 15 lines of code…</i></summary>

> ```
>     def can_take_work(self) -> bool:
> 
>         with self.lock:
> 
>             if any([
> 
>                     computation for computation in self.assigned_subtasks
> 
>                     if not computation.single_core
> 
>             ]):
> 
>                 return False
> 
>             return len(self.assigned_subtasks) < self.max_num_cores
> ```
</details>


pls consider code below instead
```
    def can_take_work(self) -> bool:
        self.free_cores > 0
```
___
*[golem/task/taskcomputer.py, line 749 at r2](https://reviewable.io/reviews/golemfactory/golem/4889#-LtjE1PPDoSffc2MqW9t:-LtjE1PPDoSffc2MqW9u:b-m97t9u) ([raw file](https://github.com/golemfactory/golem/blob/6783ae79e57b0013b8ce46534718bf9ea718f05c/golem/task/taskcomputer.py#L749)):*
I'd change to debug here. 
> ```
> logger.debug(f"num_cores={config_desc.num_cores}")
> ```
___
*[golem/task/taskcomputer.py, line 754 at r2](https://reviewable.io/reviews/golemfactory/golem/4889#-LtjEGbr6GTqsqWK8lDm:-LtjEGbr6GTqsqWK8lDn:b-4ovbu9) ([raw file](https://github.com/golemfactory/golem/blob/6783ae79e57b0013b8ce46534718bf9ea718f05c/golem/task/taskcomputer.py#L754)):*
> ```Python
> 
>         dm = self.docker_manager
>         assert isinstance(dm, DockerManager)
> ```

seems to be a not necessary check... line below will fail if dm is not an instance of `DockerManager` class

maybe it was added to just satisfy some linter
___
*[golem/task/taskcomputer.py, line 803 at r2](https://reviewable.io/reviews/golemfactory/golem/4889#-LtjN7wA0YRzqDM97BS0:-LtjN7wA0YRzqDM97BS1:b3m0n2g) ([raw file](https://github.com/golemfactory/golem/blob/6783ae79e57b0013b8ce46534718bf9ea718f05c/golem/task/taskcomputer.py#L803)):*
> ```Python
>             if not [
>                     c for c in self.assigned_subtasks
>                     if c.assigned_subtask_id == task_id
> ```

obviously should be corrected

**Comment by tworec** [comment]
___
*[golem/task/taskserver.py, line 491 at r2](https://reviewable.io/reviews/golemfactory/golem/4889#-LtjRjrg0POPTB3no_10:-LtjRjrg0POPTB3no_11:bdt63c6) ([raw file](https://github.com/golemfactory/golem/blob/6783ae79e57b0013b8ce46534718bf9ea718f05c/golem/task/taskserver.py#L491)):*
> ```
> logger.error("Trying to assign a task, but we can not take more work at the moment")
> ```

File: golem/task/server/resources.py [hunk 1]
```diff
@@ -33,7 +33,7 @@
 class TaskResourcesMixin:
     """Resource management functionality of TaskServer"""
 
-    HANDSHAKE_TIMEOUT = 20  # s
+    HANDSHAKE_TIMEOUT = 60  # s
```
Comment (shadeofblue) [inline golem/task/server/resources.py]: does this change pertain to the functionality included in the pull request or is it here incidentally?

**Review by tworec — CHANGES_REQUESTED** [review]
fix marked condition

File: golem/task/taskcomputer.py [hunk 2]
```diff
@@ -705,31 +616,231 @@ def start_computation(self) -> None:  # pylint: disable=too-many-locals
...
             min_performance=ctd['performance'],
         )
-
         with self.lock:
-            self.counting_thread = None
-        self.task_server.task_keeper.task_ended(ctd['task_id'])
+            task_id = ctd['task_id']
+            if not [
+                    c for c in self.assigned_subtasks
+                    if c.assigned_subtask_id == task_id
```
Comment (tworec) [inline golem/task/taskcomputer.py]: fixme

**Review by tworec — APPROVED** [review]

File: golem/task/server/resources.py (hunk 1 above)
Comment (prekucki) [reply golem/task/server/resources.py]: I reverted it. there is a problem with persistent ban between nodes if network is small enough, but a longer timeout won't help with this.

File: golem/task/taskcomputer.py [hunk 3]
```diff
@@ -234,12 +253,10 @@ class AssignedTask:
...
-    def __init__(
-            self,
-            env_manager: EnvironmentManager,
-            work_dir: Path,
-            stats_keeper: Optional[IntStatsKeeper] = None
-    ) -> None:
+    def __init__(self,
+                 env_manager: EnvironmentManager,
+                 work_dir: Path,
+                 stats_keeper: Optional[IntStatsKeeper] = None) -> None:
```
Comment (Wiezzel) [inline golem/task/taskcomputer.py:259]: Bad indentation.

File: golem/task/taskcomputer.py [hunk 4]
```diff
@@ -268,11 +285,8 @@ def get_subtask_inputs_dir(self) -> Path:
     def _is_computing(self) -> bool:
         return self._computation is not None
 
-    def task_given(
-            self,
-            task_header: TaskHeader,
-            compute_task_def: ComputeTaskDef
-    ) -> None:
+    def task_given(self, task_header: TaskHeader,
+                   compute_task_def: ComputeTaskDef) -> None:
```
Comment (Wiezzel) [inline golem/task/taskcomputer.py:289]: Bad indentation.

File: golem/task/taskcomputer.py [hunk 5]
```diff
@@ -337,33 +348,24 @@ def _wait_until_computation_ends(self) -> defer.Deferred:
         success = False
         try:
             output_file = yield self._computation
-            logger.info(
-                'Task computation succeeded. task_id=%r subtask_id=%r',
-                assigned_task.task_id,
-                assigned_task.subtask_id
-            )
+            logger.info('Task computation succeeded. task_id=%r subtask_id=%r',
+                        assigned_task.task_id, assigned_task.subtask_id)
```
Comment (Wiezzel) [inline golem/task/taskcomputer.py:352]: Bad indentation.

File: golem/task/taskcomputer.py [hunk 6]
```diff
@@ -403,23 +405,16 @@ def get_current_computing_env(self) -> Optional[EnvId]:
             return None
         return self._assigned_task.env_id
 
-    def change_config(
-            self,
-            config_desc: ClientConfigDescriptor,
-            work_dir: Path
-    ) -> defer.Deferred:
+    def change_config(self, config_desc: ClientConfigDescriptor,
+                      work_dir: Path) -> defer.Deferred:
```
Comment (Wiezzel) [inline golem/task/taskcomputer.py:409]: Bad indentation.

File: golem/task/taskcomputer.py [hunk 7]
```diff
@@ -403,23 +405,16 @@ def get_current_computing_env(self) -> Optional[EnvId]:
...
-                config_desc.max_memory_size,
-                unit=MemSize.kibi,
-                to_unit=MemSize.mebi
-            )
-        )
+        config_dict = dict(work_dirs=[work_dir],
+                           cpu_count=config_desc.num_cores,
+                           memory_mb=scale_memory(config_desc.max_memory_size,
+                                                  unit=MemSize.kibi,
+                                                  to_unit=MemSize.mebi))
```
Comment (Wiezzel) [inline golem/task/taskcomputer.py:417]: Bad indentation.

File: golem/task/taskcomputer.py [hunk 8]
```diff
@@ -527,29 +517,22 @@ def task_computed(self, task_thread: TaskThread) -> None:
...
-                self.task_server.send_task_failed(
-                    subtask_id,
-                    subtask['task_id'],
-                    task_thread.error_msg,
-                    reason
-                )
-
+                stats.increase_stat('tasks_with_errors')
+                task_server.send_task_failed(subtask_id, subtask['task_id'],
+                                             task_thread.error_msg, reason)
```
Comment (Wiezzel) [inline golem/task/taskcomputer.py:527]: Bad indentation.

**Review by Wiezzel — CHANGES_REQUESTED** [review]
Please do not break indentation. Our [style guide](https://github.com/golemfactory/styleguide/blob/gh-pages/pyguide.md#34-indentation) explicitly says not to use indentation dependent on names and go for multiples of 4 instead. There are some changes in this PR that don't seem to have any other purpose than changing indentation from correct to incorrect. I marked out a few but soon gave up because there were too many.

File: golem/task/taskserver.py [hunk 9]
```diff
@@ -333,56 +331,57 @@ def _request_random_task(self) -> None:
...
+
+            if (self.client.concent_service.enabled and
+                    self.client.concent_service.required_as_provider and
+                    not theader.concent_enabled):
                 supported = supported.join(
                     SupportStatus.err({
                         UnsupportReason.CONCENT_REQUIRED: True,
-                    }),
-                )
+                    }),)
```
Comment (shadeofblue) [inline golem/task/taskserver.py:376]: @prekucki I honestly don't understand the need to update the code formatting for the sake of code formatting itself.... isn't this pull request complicated enough to maybe abstain from meddling with formatting at the same time?

especially if the changes you've made are, arguably, for worse, not better? ...

File: golem/task/taskserver.py [hunk 10]
```diff
@@ -895,17 +884,19 @@ def income_listener(self, event='default', node_id=None, **kwargs):
...
 
-    def finished_subtask_listener(self,  # pylint: disable=too-many-arguments
-                                  event='default', subtask_id=None,
-                                  min_performance=None, **_kwargs):
+    def finished_subtask_listener(
+            self,  # pylint: disable=too-many-arguments
+            event='default',
+            subtask_id=None,
+            min_performance=None,
+            **_kwargs):
```
Comment (shadeofblue) [inline golem/task/taskserver.py:892]: especially here, when you're doing the opposite change ... what's the point?

**Review by shadeofblue — CHANGES_REQUESTED** [review]
please revert the formatting-only changes in this pull request... it's not a place or time to do them in a pull request that's sufficiently complicated even without them

**Comment by shadeofblue** [comment]
> Please do not break indentation. Our [style guide](https://github.com/golemfactory/styleguide/blob/gh-pages/pyguide.md#34-indentation) explicitly says not to use indentation dependent on names and go for multiples of 4 instead. There are some changes in this PR that don't seem to have any other purpose than changing indentation from correct to incorrect. I marked out a few but soon gave up because there were too many.

_exactly_ my impression ... :( ...

**Pull request closed without merging**